pallets/investments/src/lib.rs

```diff
@@ -0,0 +1,887 @@
+		///
+		/// The type of collection will always be `CollectType::Closing`.
+		#[pallet::weight(0)]
+		pub fn collect_for(
+			origin: OriginFor<T>,
+			who: T::AccountId,
+			asset_id: T::AssetId,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+
+			Self::do_collect(who, asset_id, CollectType::Closing)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T>
+where
+	<T::Accountant as AssetAccountant<T::AccountId>>::AssetInfo:
+		AssetProperties<T::AccountId, Currency = CurrencyOf<T>>,
+{
+	fn get_order(
+		orders: &mut BoundedVec<OrderOf<T>, T::MaxOutstandingCollects>,
+		curr_id: T::OrderId,
+	) -> Result<&mut OrderOf<T>, DispatchError> {
+		let is_latest = orders
+			.get_mut(orders.len())
+			.map(|order| order.id == curr_id)
+			.unwrap_or(false);
+
+		if !is_latest {
+			ensure!(
+				orders
+					.try_push(Order {
+						invest: T::Amount::zero(),
+						redeem: T::Amount::zero(),
+						id: curr_id,
+					})
+					.is_ok(),
+				Error::<T>::CollectRequired
+			);
+		}
+
+		Ok(orders
+			.get_mut(orders.len())
+			.expect("Latest element is current order and is Some. qed."))
+	}
+
+	pub(crate) fn do_collect(
+		who: T::AccountId,
+		asset_id: T::AssetId,
+		collect_type: CollectType,
+	) -> DispatchResultWithPostInfo {
+		let info = T::Accountant::info(asset_id.clone()).map_err(|_| Error::<T>::UnknownAsset)?;
+		let (collected_ids, collection) = Orders::<T>::try_mutate(
+			&who,
+			&asset_id,
+			|orders| -> Result<(Vec<T::OrderId>, Collection<T::Amount>), DispatchError> {
+				let mut collection = Collection::<T::Amount>::default();
+				let mut collected = Vec::with_capacity(orders.len());
+
+				for order in orders.iter() {
+					// We subjectively decide here, that it is better not to error out if in the sequence of orders
+					// one has not yet been cleared.
+					//
+					// Instead we signal the collected orders via the vector if OrderId's in the event.
+					if let Ok(fulfillment) = ClearedOrders::<T>::try_get(asset_id.clone(), order.id)
+					{
+						Pallet::<T>::acc_payout_invest(&mut collection, &fulfillment, &order)?;
+						Pallet::<T>::acc_payout_redeem(&mut collection, &fulfillment, &order)?;
+						Pallet::<T>::acc_remaining_invest(&mut collection, &fulfillment, &order)?;
+						Pallet::<T>::acc_remaining_invest(&mut collection, &fulfillment, &order)?;
+						collected.push(order.id);
+					};
+				}
+
+				// Transfer collected amounts from investment and redemption
+				let asset_account = AssetAccount {
+					asset_id: asset_id.clone(),
+				}
+				.into_account_truncating();
+				T::Tokens::transfer(
+					info.denomination_currency(),
+					&asset_account,
+					&who,
+					collection.payout_asset_invest,
+					false,
+				)?;
+				T::Tokens::transfer(
+					info.payment_currency(),
+					&asset_account,
+					&who,
+					collection.payout_asset_redeem,
+					false,
+				)?;
+
+				// drain the orders that have been collected
+				orders.retain(|order| !collected.contains(&order.id));
```

Comment:
Technically this is O(n^2) worst case, if we collect all orders.

I think the for loop above could be turned into the retain call, and we do the collection and removal all in one place?

Reply:
I logically made it like this. But because rust can not return early from these iterators it is still a for loop but O(2*n) now.